### ISSUE TITLE
F821 false positive in notebook when using magic command as variables

### ISSUE THREAD

**Issue by mathause** [astral-sh/ruff#9648]
ruff gives false positives for F821 (Undefined name) in notebooks when using magic command as variables. Ok, this is _very_ specific and potentially related to #9644.

In a jupyter notebook when defining a variable which has the same name as a ipython magic command and using this variable in a following cell and not as the first command triggers F821 (wrongly).

Only classify magic commands as magic if they are preceded with a `%`?

**command**

```bash
ruff check --no-cache --select F821 test.ipynb
```

**version**

```
ruff --version
ruff 0.1.14
```

**MVCE**

(I am not sure how to create a MVCE for a jupyter notebook, but I guess it's easy enough...)

![image](https://github.com/astral-sh/ruff/assets/10194086/b81030dd-fd97-4369-912b-b336b699923b)

**Comment by charliermarsh** [comment]
Yeah we can probably improve our heuristics here to specifically detect assignments. The automagics are actually really complicated because it means that (in general) cells may or may not be unsafe depending on the order of their execution (which we can't know). For example: https://github.com/astral-sh/ruff/pull/8398#issuecomment-1790345709.

**Comment by charliermarsh** [comment]
I'll try to fix this specific case today :)

**Comment by charliermarsh** [comment]
But I need to audit the list of automagics which I dread :joy:

**Comment by mathause** [comment]
Thanks!

**Comment by charliermarsh** [comment]
Of course! If you see more of these, let us know, since we have to rely on some heuristics for this given the flexibility of notebooks.
